catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx

```diff
@@ -86,32 +109,108 @@ export type FilesAction = tagged.InstanceOf<typeof FilesAction>
 
 export type LocalFile = FileWithPath & FileWithHash
 
+// TODO: queue/array
+interface ChangedDict {
+  logicalKey?: string
+  // TODO: meta?: Types.JsonRecord
+}
+
+type AddedFile = (LocalFile | Model.S3File) & {
+  changed?: ChangedDict
+  conflict?: string
+}
+
+type ExistingFile = Model.PackageEntry & {
+  changed?: ChangedDict
+  conflict?: string
+}
+
+type AnyFile = ExistingFile | AddedFile
+
+type AddedItems = Record<string, AddedFile>
+
+type ExistingItems = Record<string, ExistingFile>
+
 export interface FilesState {
-  added: Record<string, LocalFile | Model.S3File>
+  added: AddedItems
   deleted: Record<string, true>
-  existing: Record<string, Model.PackageEntry>
+  existing: ExistingItems
   // XXX: workaround used to re-trigger validation and dependent computations
   // required due to direct mutations of File objects
   counter?: number
 }
 
-const addMetaToFile = (
-  file: Model.PackageEntry | LocalFile | Model.S3File,
-  meta?: Model.EntryMeta,
-) => {
+function cloneDomFile<F extends AnyFile>(file: F, omitProperty: string): F {
+  const fileCopy = new window.File([file as File], (file as File).name, {
+    type: (file as File).type,
+  })
+  const properties = ['conflict', 'meta', 'changed', 'hash']
+  properties.forEach((propName) => {
+    const property = Object.getOwnPropertyDescriptor(file, propName)
+    if (property?.value !== undefined && omitProperty !== propName) {
+      Object.defineProperty(fileCopy, propName, property)
+    }
+  })
+  return fileCopy as F
+}
+
+function setKeyValue<T>(key: string, value: T, file: AnyFile): AnyFile {
   if (file instanceof window.File) {
-    const fileCopy = new window.File([file as File], (file as File).name, {
-      type: (file as File).type,
-    })
-    Object.defineProperty(fileCopy, 'meta', {
-      value: meta,
-    })
-    Object.defineProperty(fileCopy, 'hash', {
-      value: (file as FileWithHash).hash,
-    })
+    const fileCopy = cloneDomFile(file, key)
+    if (value !== undefined) {
+      Object.defineProperty(fileCopy, key, {
+        value,
+      })
+    }
     return fileCopy
   }
-  return R.assoc('meta', meta, file)
+  return R.assoc(key, value, file)
+}
+
+function addFile<T extends AnyFile, M extends Record<string, AnyFile>>(
+  path: string,
+  file: T,
+  mainItems: M,
```

Comment:
why do you need this type var here? you're not using it. or is it needed for the return type?

Reply:
Yes, I can skip it now. Earlier types were stricter, but I gave up